rust/src/sans_io/read.rs

```diff
@@ -0,0 +1,894 @@
+    compressed_data: Vec<u8>,
+    data_section_hasher: Option<crc32fast::Hasher>,
+    calculated_data_section_crc: Option<u32>,
+    decompressors: HashMap<String, Box<dyn Decompressor>>,
+    options: LinearReaderOptions,
+    at_eof: bool,
+    failed: bool,
+}
+
+impl LinearReader {
+    pub fn new() -> Self {
+        Self::new_with_options(LinearReaderOptions::default())
+    }
+
+    pub fn new_with_options(options: LinearReaderOptions) -> Self {
+        LinearReader {
+            currently_reading: if options.skip_start_magic {
+                RecordOpcodeLength
+            } else {
+                StartMagic
+            },
+            from: File,
+            uncompressed_data: Vec::new(),
+            uncompressed_data_start: 0,
+            uncompressed_data_end: 0,
+            compressed_data: Vec::new(),
+            compressed_data_start: 0,
+            compressed_data_end: 0,
+            data_section_hasher: match options.data_section_crc_validation_strategy {
+                CRCValidationStrategy::None => None,
+                CRCValidationStrategy::BeforeReading => {
+                    panic!("data section crc validation before reading not supported");
+                }
+                CRCValidationStrategy::AfterReading => Some(crc32fast::Hasher::new()),
+            },
+            calculated_data_section_crc: None,
+            decompressors: HashMap::new(),
+            at_eof: false,
+            options,
+            failed: false,
+        }
+    }
+
+    /// Constructs a linear reader that will iterate through all records in a chunk.
+    pub(crate) fn for_chunk(header: ChunkHeader) -> McapResult<Self> {
+        let mut result = Self::new_with_options(LinearReaderOptions {
+            skip_end_magic: true,
+            skip_start_magic: true,
+            chunk_crc_validation_strategy: CRCValidationStrategy::AfterReading,
+            data_section_crc_validation_strategy: CRCValidationStrategy::None,
+            emit_chunks: false,
+        });
+        result.currently_reading = RecordOpcodeLength;
+        result.from = Chunk(ChunkState {
+            decompressor: result.get_decompressor(&header.compression)?,
+            hasher: Some(crc32fast::Hasher::new()),
+            crc: header.uncompressed_crc,
+            next_read_size: DEFAULT_CHUNK_DATA_READ_SIZE,
+            compressed_remaining: header.compressed_size,
+            padding_after_compressed_data: 0,
+        });
+        Ok(result)
+    }
+
+    fn get_decompressor(&mut self, name: &str) -> McapResult<Box<dyn Decompressor>> {
+        if let Some(decompressor) = self.decompressors.remove(name) {
+            return Ok(decompressor);
+        }
+        match name {
+            #[cfg(feature = "zstd")]
+            "zstd" => Ok(Box::new(zstd::ZstdDecoder::new())),
+            #[cfg(feature = "lz4")]
+            "lz4" => Ok(Box::new(lz4::Lz4Decoder::new()?)),
+            "" => Ok(Box::new(NoneDecompressor {})),
```

Comment:
Does it make sense to have a NoneDecompressor instead of using Optional<None> here?

The latter seems like it would be lower overhead, no need to allocate memory or go through the dynamic method dispatch.

*edit*: it looks like NoneDecompressor copies from one buffer to the other, and you'd have to otherwise inline that code after checking for None. I guess this is fine.